app/src/community/java/com/twilio/video/app/auth/CommunityAuthenticator.kt

```diff
@@ -17,39 +17,56 @@
 package com.twilio.video.app.auth
 
 import android.content.SharedPreferences
-import com.twilio.video.app.auth.LoginEvent.CommunityLoginEvent
+import com.twilio.video.app.auth.LoginResult.CommunityLoginFailureResult
 import com.twilio.video.app.auth.LoginResult.CommunityLoginSuccessResult
-import com.twilio.video.app.data.Preferences
+import com.twilio.video.app.data.PASSCODE
+import com.twilio.video.app.data.Preferences.DISPLAY_NAME
+import com.twilio.video.app.data.api.TokenService
+import com.twilio.video.app.util.putString
+import com.twilio.video.app.util.remove
 import io.reactivex.Observable
-import io.reactivex.Single
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.rx2.rxSingle
 import timber.log.Timber
+import kotlin.coroutines.CoroutineContext
 
 class CommunityAuthenticator constructor(
-    private val preferences: SharedPreferences
+    private val sharedPreferences: SharedPreferences,
+    private val tokenService: TokenService,
+    private val coroutineContext: CoroutineContext = Dispatchers.IO
 ) : Authenticator {
 
     override fun login(loginEventObservable: Observable<LoginEvent>): Observable<LoginResult> {
-        return Single.create<LoginResult> { observable ->
-            loginEventObservable.subscribe({ loginEvent ->
-                if (loginEvent is CommunityLoginEvent) {
-                    preferences.edit().putString(Preferences.DISPLAY_NAME, loginEvent.displayName).apply()
-                    observable.onSuccess(CommunityLoginSuccessResult)
-                }
-            }, {
-                Timber.e(it)
-            })
-        }.toObservable()
+        return Observable.empty()
     }
 
     override fun login(loginEvent: LoginEvent): Observable<LoginResult> {
-        return login(Observable.just(loginEvent))
+        return rxSingle(coroutineContext) {
+            if (loginEvent is LoginEvent.CommunityLoginEvent) {
+                sharedPreferences.putString(DISPLAY_NAME, loginEvent.identity)
+                sharedPreferences.putString(PASSCODE, loginEvent.passcode) // TODO Encrypt
+
+                try {
+                    tokenService.getToken(loginEvent.identity)
+                    CommunityLoginSuccessResult
+                } catch (e: Exception) {
```

Comment:
This just catches any error returned from the service now. I'm currently working on deserializing the specific error response from the server and mapping it to an enum type that the UI can key off of.